bolt-sidecar/src/api/commitments/firewall/receiver.rs

```diff
@@ -0,0 +1,616 @@
+            let shutdown_rx = shutdown_rx.resubscribe();
+            let signer = signer.clone();
+
+            tokio::spawn(async move {
+                retry_with_backoff_if(
+                    MAX_RETRIES,
+                    Some(retry_config),
+                    // NOTE: this needs to be a closure because it must be re-called upon failure.
+                    // As such we also need to clone the inputs again.
+                    move || {
+                        let url = url.clone();
+                        let jwt = ProposerAuthClaims::new_from_signer(
+                            url.clone(),
+                            self.chain,
+                            None,
+                            signer.clone(),
+                        )
+                        .to_jwt()
+                        .expect("failed to produce JWT");
+
+                        let api_events_tx = api_events_tx.clone();
+                        let ping_rx = ping_rx.resubscribe();
+                        let shutdown_rx = shutdown_rx.resubscribe();
+
+                        async move {
+                            handle_connection(url, state, jwt, api_events_tx, ping_rx, shutdown_rx)
+                                .await
+                        }
+                    },
+                    handle_error,
+                )
+                .await
+            });
+        }
+
+        api_events_rx
+    }
+
+    /// Returns the local addr the server is listening on (or configured with).
+    pub fn urls(&self) -> &[Url] {
+        self.urls.as_slice()
+    }
+}
+
+/// Opens the websocket connection and starts the commitment request processor.
+async fn handle_connection(
+    url: String,
+    state: ProcessorState,
+    jwt: String,
+    api_events_tx: mpsc::Sender<CommitmentEvent>,
+    ping_rx: broadcast::Receiver<()>,
+    shutdown_rx: broadcast::Receiver<()>,
+) -> Result<(), ConnectionHandlerError> {
+    let ws_config =
+        WebSocketConfig { max_message_size: Some(MAX_MESSAGE_SIZE), ..Default::default() };
+
+    let mut request = url.clone().into_client_request()?;
+    request
+        .headers_mut()
+        .insert("Authorization", format!("Bearer {}", jwt).parse().expect("valid header"));
+
+    match connect_async_with_config(request, Some(ws_config), USE_NAGLE).await {
+        Ok((stream, response)) => {
+            info!(?url, ?response, "opened websocket connection");
+            let (write_sink, read_stream) = stream.split();
+
+            // For each opened connection, create a new commitment processor
+            // able to handle incoming message requests.
+            let commitment_request_processor = CommitmentRequestProcessor::new(
+                url,
+                state,
+                api_events_tx.clone(),
+                write_sink,
+                read_stream,
+                ping_rx.resubscribe(),
+                shutdown_rx.resubscribe(),
+            );
```

Comment:
nit: why not `.split()` the stream in `CommitmentRequestProcessor::new` instead of here?

Reply:
Good idea!